vllm/entrypoints/llm.py

```diff
@@ -734,10 +740,29 @@ def _run_engine(
                                 f"est. speed input: {in_spd:.2f} toks/s, "
                                 f"output: {out_spd:.2f} toks/s")
                         pbar.update(1)
+            if self.llm_engine.cache_service_metrics is not None:
```

Comment:
The metrics calculation and logging logic should go inside the individual metrics class instead of embedded into the vLLM engine high-level event-loop:
- the engine logic should encodes abstract operations such as collecting one measurement per step, or finish the measurements (and trigger all derived metrics calculation) once a request is finished.
- the individual metrics class will handle its own logic of which measurement to track and derived metrics to calculate and log.

This way, the engine code will remain stable while we plug in different metrics implementations